source/ShipManager.cpp

```diff
@@ -0,0 +1,190 @@
+			else if(key == "requires outfits")
+				requiresOutfits = true;
+			else
+				child.PrintTrace("Error: Skipping unrecognized token.");
+		}
+		else
+			child.PrintTrace("Error: Skipping unrecognized token.");
+	}
+
+	if(taking && !id.empty() && amount > 1)
+		node.PrintTrace("Error: Use of \"id\" to refer to the ship is only supported when \"amount\" is equal to 1.");
+}
+
+
+
+void ShipManager::Save(DataWriter &out) const
+{
+	out.Write(Giving() ? "give" : "take", "ship", model->VariantName(), name);
+	out.BeginChild();
+	{
+		out.Write("amount", amount);
+		if(!id.empty())
+			out.Write("id", id);
+		if(unconstrained)
+			out.Write("unconstrained");
+		if(takeOutfits)
+			out.Write("with outfits");
+	}
+	out.EndChild();
+}
+
+
+
+bool ShipManager::CanBeDone(const PlayerInfo &player) const
+{
+	// If we are giving ships there are no conditions to meet.
+	return Giving() || static_cast<int>(SatisfyingShips(player).size()) == amount;
+}
+
+
+
+void ShipManager::Do(PlayerInfo &player) const
+{
+	if(model->ModelName().empty())
+		return;
+
+	string shipName;
+	if(Giving())
+	{
+		for(int i = 0; i < amount; ++i)
+			shipName = player.GiftShip(model, name, id)->Name();
+	}
+	else
+	{
+		auto toTake = SatisfyingShips(player);
+		if(toTake.size() == 1)
+			shipName = toTake.begin()->get()->Name();
+		for(const auto &ship : toTake)
+			player.TakeShip(ship.get(), model, takeOutfits);
+	}
+	Messages::Add((amount == 1 ? "The " + model->ModelName() + " \"" + shipName + "\" was " :
+		to_string(amount) + " " + model->PluralModelName() + " were ") +
+		(Giving() ? "added to" : "removed from") + " your fleet.", Messages::Importance::High);
+}
+
+
+
+const Ship *ShipManager::ShipModel() const
+{
+	return model;
+}
+
+
+
+const string &ShipManager::Id() const
+{
+	return id;
+}
+
+
+
+bool ShipManager::Giving() const
+{
+	return !taking;
+}
+
+
+
+vector<shared_ptr<Ship>> ShipManager::SatisfyingShips(const PlayerInfo &player) const
+{
+	const System *here = player.GetSystem();
+	const auto shipToTakeId = player.GiftedShips().find(id);
+	bool foundShip = shipToTakeId != player.GiftedShips().end();
+	vector<shared_ptr<Ship>> satisfyingShips;
+
+	for(const auto &ship : player.Ships())
+		if((ship->ModelName() == model->ModelName())
+			&& (unconstrained || (ship->GetSystem() == here && !ship->IsDisabled() && !ship->IsParked()))
+			&& (id.empty() || (foundShip && ship->UUID() == shipToTakeId->second))
+			&& (name.empty() || name == ship->Name()))
+		{
+			bool hasRequiredOutfits = true;
+			// If "take outfits" or "requires outfits" is specified,
+			// this ship must have each outfit specified in that variant definition.
+			if(requiresOutfits || takeOutfits)
+				for(const auto &it : model->Outfits())
+				{
+					const auto &outfit = ship->Outfits().find(it.first);
+					int amountEquipped = (outfit != ship->Outfits().end() ? outfit->second : 0);
+					if(it.second > amountEquipped)
+					{
+						hasRequiredOutfits = false;
+						continue;
+					}
+				}
```

Comment:
An O(n) approach instead of O(n*log(n)), I think.
(Really, O(N+n) instead of O(N*log(n)), where `N` is the size of the `model` outfit map and `n` is the size of the target ship outfit map.
```suggestion
			if(requiresOutfits || takeOutfits)
			{
				auto modelIt = model->Outfits().begin();
				auto targetIt = ship->Outfits().begin();
				const auto &modelEnd = model->Outfits().end();
				const auto &targetEnd = ship->Outfits().end();
				while(modelIt != modelEnd && targetIt != targetEnd)
				{
					if(targetIt->first != modelIt->first)
					{
						++targetIt;
						continue;
					}
					else if(targetIt->second < modelIt->second)
					{
						++targetIt;
						++modelIt;
					}
				}
				if(targetIt == targetEnd && modelIt != modelEnd)
					hasRequiredOutfits = false;
			}
```

Reply:
This seems complicated I'll look at it shortly on my computer ~~dont worry I get the n calculations~~

Reply:
is it worth making it requiring a computer science bachelor to understand? I think that since it only happens on landing its fine as is honestly
worst time it would happen once per system entry, still fine